infrastructure/src/test/java/org/corfudb/infrastructure/LogReplicationServerTest.java

```diff
@@ -126,19 +150,63 @@ public void testHandleLeadershipQuery() {
+                0L, SAMPLE_HOSTNAME, false, mock(LogReplicationPluginConfig.class));
+        lrServer = spy(new LogReplicationServer(context, new HashSet<>(), topology, metadataManager,
+                replicationContext));
+        lrServer.getHandlerMethods().handle(request, null, mockServerRouter);
         argument = ArgumentCaptor.forClass(ResponseMsg.class);
-        verify(mockServerRouter, atMost(2))
-            .sendResponse(argument.capture());
+        verify(mockServerRouter, atMost(2)).sendResponse(argument.capture());
+        Assertions.assertThat(argument.getValue().getPayload().getLrLeadershipResponse().getIsLeader()).isFalse();
+
+        // verify the response when not the leader
+        replicationContext = new LogReplicationContext(mock(LogReplicationConfigManager.class),
+                0L, SAMPLE_HOSTNAME, false, mock(LogReplicationPluginConfig.class));
+        lrServer = spy(new LogReplicationServer(context, new HashSet<>(), topology, metadataManager,
+                replicationContext));
+        lrServer.getHandlerMethods().handle(request, null, mockServerRouter);
+        argument = ArgumentCaptor.forClass(ResponseMsg.class);
+        verify(mockServerRouter, atMost(3)).sendResponse(argument.capture());
+        Assertions.assertThat(argument.getValue().getPayload().getLrLeadershipResponse().getIsLeader()).isFalse();
+    }
+
+    @Test
+    public void testHandleLeadershipQueryOnMixedVersion() {
+        final LogReplicationLeadershipRequestMsg leadershipQuery =
+                LogReplicationLeadershipRequestMsg.newBuilder().build();
+        final RequestMsg request =
+                getRequestMsg(HeaderMsg.newBuilder()
+                        .setVersion(getDefaultProtocolVersionMsg())
+                        .setIgnoreClusterId(true)
+                        .setIgnoreEpoch(true)
+                        .setClientId(getUuidMsg(UUID.randomUUID()))
+                                .build(),
+                        CorfuMessage.RequestPayloadMsg.newBuilder()
+                                .setLrLeadershipQuery(leadershipQuery).build());
+
+        doReturn(SAMPLE_HOSTNAME).when(context).getLocalEndpoint();
+
+        // when node is the leader and when sessionManger knows about the session
+        lrServer.getHandlerMethods().handle(request, null, mockServerRouter);
+        ArgumentCaptor<ResponseMsg> argument = ArgumentCaptor.forClass(ResponseMsg.class);
+        verify(mockServerRouter, atMost(1)).sendResponse(argument.capture());
         Assertions.assertThat(argument.getValue().getPayload()
-            .getLrLeadershipResponse().getIsLeader()).isTrue();
+                .getLrLeadershipResponse().getIsLeader()).isTrue();
+
+        //when node is the leader and sessionManager does not know about the session
+        lrServer = spy(new LogReplicationServer(context, new HashSet<>(), topology, metadataManager, replicationContext));
```

Comment:
same here.. shouldnt the topology be mocked differently?

Also, isnt this test having the same verification as testHandleLeadershipQuery()?

clusterId was added in the current version.  To test receipt of a request from an older version, can you also write a test where the cluster id is not set in the request?

Reply:
No, this is to test the race when the sink received the new topology, but the session manager has not yet created the session, and the sink received a request from the source.

Its having the same verification but its with a request from LR V1 where the clusterId is not present

Reply:
ok, got it.  testLeadershipQuery does not simulate the case of cluster id being absent in the request.  Can you add a comment on the test explaining in brief what it does and also that it tests these 2 cases - topology with LR server has and does not have the FULL_TABLE session?